Fails to Compile Against Graphviz 13.x

Graphviz 13.0, released in June, had the following listed in its [changelog](https://gitlab.com/graphviz/graphviz/-/blob/main/CHANGELOG.md?ref_type=heads#1300--2025-06-08):

- Breaking: gvRenderData takes its length parameter as a size_t *. The GVJ_t fields output_data_allocated and output_data_position have type size_t.

When graphviz was updated to 13.x in my distro's repo, it broke libmediainfo when compiled with graphviz support, which is a package I maintain for that repo. When trying to compile against the newer graphviz, the error is:

```
../../../Source/MediaInfo/Export/Export_Graph.cpp:160:60: error: cannot convert ‘unsigned int*’ to ‘size_t*’ {aka ‘long unsigned int*’}
  160 |         gvRenderData(Context, Graph, "svg:cairo", &Buffer, &Size);
      |                                                            ^~~~~
      |                                                            |
      |                                                            unsigned int*
In file included from ../../../Source/MediaInfo/Export/Export_Graph.cpp:35:
/usr/include/graphviz/gvc.h:99:49: note:   initializing argument 5 of ‘int gvRenderData(GVC_t*, graph_t*, const char*, char**, size_t*)’
   99 |                          char **result, size_t *length);
      |                                         ~~~~~~~~^~~~~~
../../../Source/MediaInfo/Export/Export_Graph.cpp:162:54: error: cannot convert ‘unsigned int*’ to ‘size_t*’ {aka ‘long unsigned int*’}
  162 |         gvRenderData(Context, Graph, "svg", &Buffer, &Size);
      |                                                      ^~~~~
      |                                                      |
      |                                                      unsigned int*
In file included from ../../../Source/MediaInfo/Export/Export_Graph.cpp:35:
/usr/include/graphviz/gvc.h:99:49: note:   initializing argument 5 of ‘int gvRenderData(GVC_t*, graph_t*, const char*, char**, size_t*)’
   99 |                          char **result, size_t *length);
      |                                         ~~~~~~~~^~~~~~
```

We are able to fix this locally with the following sed command:

    sed -i 's|unsigned int|size_t|' Source/MediaInfo/Export/Export_Graph*

However, this is not backwards compatible with graphviz 12.x and earlier, so it would seem the code would need some sort of if/then statement to determine the graphviz version and use the correct parameters, but that is beyond my programming expertise.

--- a/.github/workflows/MediaInfoLib_Checks.yml
+++ b/.github/workflows/MediaInfoLib_Checks.yml
@@ -36,6 +36,10 @@ jobs:
           - os: windows-latest
             environement: minimal-with-xml
         include:
+          - os: ubuntu-latest
+            environement: link-curl-graphviz
+          - os: macos-latest
+            environement: link-curl-graphviz
           - environement: gcc
             CXXFLAGS: ""
             RUN_TESTS: "all"
@@ -67,10 +71,10 @@ jobs:
         run: |
           if [ "$RUNNER_OS" == "Linux" ]; then
             sudo apt-get update -y
-            sudo apt-get install -y libxml2-utils zlib1g-dev libcurl4-gnutls-dev libmms-dev ${{ matrix.environement == 'cmake-dll' && 'g++-mingw-w64-x86-64' || '' }}
+            sudo apt-get install -y libxml2-utils zlib1g-dev libcurl4-gnutls-dev libmms-dev ${{ matrix.environement == 'cmake-dll' && 'g++-mingw-w64-x86-64' || '' }} ${{ matrix.environement == 'link-curl-graphviz' && 'libgraphviz-dev' || '' }}
           fi
           if [ "$RUNNER_OS" == "macOS" ]; then
-            brew install libtool automake
+            brew install libtool automake ${{ matrix.environement == 'link-curl-graphviz' && 'graphviz' || '' }}
           fi
       - name: zlib
         if: ${{ (runner.os == 'Windows') || matrix.environement == 'cmake-dll' }}
@@ -89,7 +93,7 @@ jobs:
           popd
           cd Project/GNU/Library
           autoreconf -if
-          ./configure --enable-silent-rules --with-libcurl=runtime
+          ./configure --enable-silent-rules ${{ matrix.environement == 'link-curl-graphviz' && '--with-libcurl --with-graphviz' || '--with-libcurl=runtime --with-graphviz=runtime' }}
       - name: Build
         if: ${{ (matrix.environement != 'cmake') && (matrix.environement != 'cmake-dll') }}
         run: |

--- a/Project/GNU/Library/configure.ac
+++ b/Project/GNU/Library/configure.ac
@@ -558,6 +558,38 @@ else
 fi
 
 dnl -------------------------------------------------------------------------
+dnl Graphviz version detection
+
+if test "$using_graphviz" != "no" -a "$using_graphviz" != ""; then
+	# Try to detect Graphviz version using pkg-config
+	if ${PKG_CONFIG:-pkg-config} --exists libgvc; then
+		GV_PKG_VERSION=$(${PKG_CONFIG:-pkg-config} --modversion libgvc)
+		AC_MSG_NOTICE([Detected Graphviz version via pkg-config: $GV_PKG_VERSION])
+	else
+		# Fallback: use $(dot -V)
+		AC_PATH_PROG([DOT], [dot], [no])
+		if test "$DOT" != "no"; then
+			GV_PKG_VERSION=$($DOT -V 2>&1 | grep -o '[0-9]\+\.[0-9]\+\.[0-9]\+' | head -n1)
+			AC_MSG_NOTICE([Detected Graphviz version via dot -V: $GV_PKG_VERSION])
+		else
+			AC_MSG_WARN([Could not determine Graphviz version - neither pkg-config nor dot found])
+		fi
+	fi
+
+	# Extract major version
+	if test -n "$GV_PKG_VERSION"; then
+		GV_VERSION_MAJOR=$(echo "$GV_PKG_VERSION" | cut -d. -f1)
+		AC_MSG_NOTICE([Graphviz major version: $GV_VERSION_MAJOR])
+	else
+		GV_VERSION_MAJOR=0
+		AC_MSG_WARN([Graphviz version unknown, defaulting GV_VERSION_MAJOR=0])
+	fi
+
+	# Define GV_VERSION_MAJOR
+	AC_DEFINE_UNQUOTED([GV_VERSION_MAJOR], [$GV_VERSION_MAJOR], [Graphviz major version])
+fi
+
+dnl -------------------------------------------------------------------------
 dnl libcurl
 dnl
 if test -d $with_libcurl; then

--- a/Project/GNU/Library/configure.ac
+++ b/Project/GNU/Library/configure.ac
@@ -560,7 +560,7 @@ fi
 dnl -------------------------------------------------------------------------
 dnl Graphviz version detection
 
-if test "$using_graphviz" != "no" -a "$using_graphviz" != ""; then
+if test "$using_graphviz" != "no" -a "$using_graphviz" != "system (runtime)" -a "$using_graphviz" != ""; then
 	# Try to detect Graphviz version using pkg-config
 	if ${PKG_CONFIG:-pkg-config} --exists libgvc; then
 		GV_PKG_VERSION=$(${PKG_CONFIG:-pkg-config} --modversion libgvc)
@@ -572,7 +572,40 @@ if test "$using_graphviz" != "no" -a "$using_graphviz" != ""; then
 			GV_PKG_VERSION=$($DOT -V 2>&1 | grep -o '[0-9]\+\.[0-9]\+\.[0-9]\+' | head -n1)
 			AC_MSG_NOTICE([Detected Graphviz version via dot -V: $GV_PKG_VERSION])
 		else
-			AC_MSG_WARN([Could not determine Graphviz version - neither pkg-config nor dot found])
+			AC_MSG_CHECKING([Graphviz version via gvcVersion()])
+    
+			# Save current LIBS to restore them later
+			SAVED_LIBS=$LIBS
+			LIBS="$LIBS -lgvc"
+
+			AC_RUN_IFELSE([
+				AC_LANG_PROGRAM([[
+					#include <stdio.h>
+					#include <graphviz/gvc.h>
+				]], [[
+					FILE *f = fopen("gvcVersion.out", "w");
+					GVC_t *gvc = gvContext();
+					if (gvc && f) {
+						fprintf(f, "%s", gvcVersion(gvc));
+						fclose(f);
+						gvFreeContext(gvc);
+						return 0;
+					}
+					return 1;
+				]])
+			], [
+				GV_PKG_VERSION=`cat gvcVersion.out`
+				AC_MSG_RESULT([$GV_PKG_VERSION])
+			], [
+				AC_MSG_RESULT([failed])
+				AC_MSG_WARN([Could not determine Graphviz version - neither pkg-config nor dot found and gvcVersion() failed])
+			], [
+				# Cross-compilation fallback
+				AC_MSG_RESULT([cross-compiling; cannot run gvcVersion()])
+				AC_MSG_WARN([Could not determine Graphviz version - neither pkg-config nor dot found and gvcVersion() failed])
+			])
+
+			LIBS=$SAVED_LIBS
 		fi
 	fi
 

--- a/Source/MediaInfo/Export/Export_Graph.cpp
+++ b/Source/MediaInfo/Export/Export_Graph.cpp
@@ -162,10 +162,11 @@ Ztring Dot2Svg(const Ztring& Dot)
         if (!strcmp(Renderers[Pos], "cairo"))
             Cairo=true;
 
-        free(Renderers[Pos]); // Note: Cannot free memory if Graphviz DLL uses different (debug vs release) ucrt from MediaInfo
+        gvFreeRenderData(Renderers[Pos]); // Make use of `gvFreeRenderData` to call Graphviz's internal `free` to prevent ucrt mismatch
     }
     if (Renderers_Size)
-        free(Renderers); // Note: Cannot free memory if Graphviz DLL uses different (debug vs release) ucrt from MediaInfo
+        gvFreeRenderData(reinterpret_cast<char*>(Renderers)); // Make use of `gvFreeRenderData` to call Graphviz's internal `free` to prevent ucrt mismatch
+    Renderers = nullptr;
 
     Graph=agmemread(Dot.To_UTF8().c_str()); 
     if (!Graph)

--- a/Source/MediaInfo/Export/Export_Graph.cpp
+++ b/Source/MediaInfo/Export/Export_Graph.cpp
@@ -157,7 +157,13 @@ Ztring Dot2Svg(const Ztring& Dot)
         return ToReturn;
 
     int Renderers_Size=0;
-    char** Renderers=gvPluginList(Context, "render", &Renderers_Size);
+    // Graphviz 9: Breaking: The str parameter from gvPluginList has been removed.
+    // Ubuntu etc. are still version 2 until 26.04 (R) where it is updated to version 14
+    #if !defined(MEDIAINFO_GRAPHVIZ_DLL_RUNTIME) && defined(GV_VERSION_MAJOR) && GV_VERSION_MAJOR != 0 && GV_VERSION_MAJOR < 9
+        char** Renderers=gvPluginList(Context, "render", &Renderers_Size, nullptr);
+    #else
+        char** Renderers=gvPluginList(Context, "render", &Renderers_Size);
+    #endif
     for (int Pos=0; Pos<Renderers_Size; Pos++) {
         if (!strcmp(Renderers[Pos], "cairo"))
             Cairo=true;

--- a/Source/MediaInfo/Export/Export_Graph.cpp
+++ b/Source/MediaInfo/Export/Export_Graph.cpp
@@ -110,14 +110,41 @@ Ztring Element2Html(MediaInfo_Internal &MI, stream_t StreamKind, size_t StreamPo
 //---------------------------------------------------------------------------
 
 #ifdef MEDIAINFO_GRAPHVIZ_YES
+
+// gvRenderData wrapper function to handle API changes in Graphviz version 13
+// Graphviz 13: Breaking: gvRenderData takes its length parameter as a size_t *.
+static int gvRenderData_dispatch(GVC_t* gvc, graph_t* g, const char* format, char** result, size_t* length)
+{
+    int ret{};
+    #ifdef MEDIAINFO_GRAPHVIZ_DLL_RUNTIME
+        if (atoi(gvcVersion(gvc)) >= 13) {
+            ret = ((gvRenderData_new)gvRenderData)(gvc, g, format, result, length);
+        }
+        else {
+            unsigned int uLength{};
+            ret = ((gvRenderData_old)gvRenderData)(gvc, g, format, result, &uLength);
+            *length = static_cast<size_t>(uLength);
+        }
+    #else
+        #if defined(GV_VERSION_MAJOR) && GV_VERSION_MAJOR != 0 && GV_VERSION_MAJOR < 13
+            unsigned int uLength{};
+            ret = gvRenderData(gvc, g, format, result, &uLength);
+            *length = static_cast<size_t>(uLength);
+        #else
+            ret = gvRenderData(gvc, g, format, result, length);
+        #endif
+    #endif //MEDIAINFO_GRAPHVIZ_DLL_RUNTIME
+    return ret;
+}
+
 Ztring Dot2Svg(const Ztring& Dot)
 {
     Ztring ToReturn;
-    GVC_t* Context=NULL;
-    graph_t* Graph=NULL;
-    char* Buffer=NULL;
-    unsigned int Size;
-    bool Cairo=false;
+    GVC_t* Context{};
+    graph_t* Graph{};
+    char* Buffer{};
+    size_t Size{};
+    bool Cairo{};
 
     if (!Export_Graph::Load())
         return ToReturn;
@@ -135,10 +162,10 @@ Ztring Dot2Svg(const Ztring& Dot)
         if (!strcmp(Renderers[Pos], "cairo"))
             Cairo=true;
 
-        free(Renderers[Pos]);
+        free(Renderers[Pos]); // Note: Cannot free memory if Graphviz DLL uses different (debug vs release) ucrt from MediaInfo
     }
     if (Renderers_Size)
-        free(Renderers);
+        free(Renderers); // Note: Cannot free memory if Graphviz DLL uses different (debug vs release) ucrt from MediaInfo
 
     Graph=agmemread(Dot.To_UTF8().c_str()); 
     if (!Graph)
@@ -156,10 +183,7 @@ Ztring Dot2Svg(const Ztring& Dot)
         return ToReturn;
     }
 
-    if (Cairo)
-        gvRenderData(Context, Graph, "svg:cairo", &Buffer, &Size);
-    else
-        gvRenderData(Context, Graph, "svg", &Buffer, &Size);
+    gvRenderData_dispatch(Context, Graph, Cairo ? "svg:cairo" : "svg", &Buffer, &Size);
 
     if (Buffer && Size)
         ToReturn=Ztring().From_UTF8(Buffer);
@@ -206,6 +230,7 @@ bool Export_Graph::Load()
             bool Error=false;
             ASSIGN(gvc_Module, gvContext)
             ASSIGN(gvc_Module, gvFreeContext)
+            ASSIGN(gvc_Module, gvcVersion)
             ASSIGN(gvc_Module, gvPluginList)
             ASSIGN(gvc_Module, gvLayout)
             ASSIGN(gvc_Module, gvFreeLayout)

--- a/Source/MediaInfo/Export/Export_Graph_gvc_Include.h
+++ b/Source/MediaInfo/Export/Export_Graph_gvc_Include.h
@@ -99,10 +99,13 @@ extern "C"
 // Interface
 typedef GVC_t* (*gvc_Module_gvContext)(); static gvc_Module_gvContext gvContext=NULL;
 typedef int (*gvc_Module_gvFreeContext)(GVC_t*); static gvc_Module_gvFreeContext gvFreeContext=NULL;
+typedef char* (*gvc_Module_gvcVersion)(GVC_t*); static gvc_Module_gvcVersion gvcVersion = NULL;
 typedef char** (*gvc_Module_gvPluginList)(GVC_t*, const char*, int*); static gvc_Module_gvPluginList gvPluginList=NULL;
 typedef int (*gvc_Module_gvLayout)(GVC_t*, graph_t*, const char*); static gvc_Module_gvLayout gvLayout=NULL;
 typedef int (*gvc_Module_gvFreeLayout)(GVC_t*, graph_t*); static gvc_Module_gvFreeLayout gvFreeLayout=NULL;
-typedef int (*gvc_Module_gvRenderData)(GVC_t*, graph_t*, const char*, char**, unsigned int*); static gvc_Module_gvRenderData gvRenderData=NULL;
+typedef void (*gvc_Module_gvRenderData)(); static gvc_Module_gvRenderData gvRenderData=NULL;
+typedef int (*gvRenderData_new)(GVC_t*, graph_t*, const char*, char**, size_t*);
+typedef int (*gvRenderData_old)(GVC_t*, graph_t*, const char*, char**, unsigned int*);
 typedef void (*gvc_Module_gvFreeRenderData)(char*); static gvc_Module_gvFreeRenderData gvFreeRenderData=NULL;
 typedef void (*gvc_Module_gvFinalize)(GVC_t*); static gvc_Module_gvFinalize gvFinalize=NULL;
 typedef graph_t* (*cgraph_Module_agmemread)(const char*); static cgraph_Module_agmemread agmemread=NULL;

